File: app/src/main/java/com/bopr/android/smailer/GeoLocator.java
package com.bopr.android.smailer;

import android.content.Context;
import android.location.Location;
import android.os.Looper;

import androidx.annotation.Nullable;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import static com.bopr.android.smailer.PermissionsHelper.isLocationPermissionsGranted;
import static com.google.android.gms.location.LocationRequest.PRIORITY_HIGH_ACCURACY;

/**
 * Provides last device location.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class GeoLocator {

    private static Logger log = LoggerFactory.getLogger("GeoLocator");

    private static final int DEFAULT_TIMEOUT = 2000;

    private final Context context;
    private final FusedLocationProviderClient client;
    private final Database database;

    public GeoLocator(Context context, Database database) {
        this.context = context;
        this.database = database;
        client = LocationServices.getFusedLocationProviderClient(context);
    }

    /**
     * Retrieve device location.
     *
     * @return last known device location or null when it cannot be found
     */
    @Nullable
    public GeoCoordinates getLocation() {
        return getLocation(DEFAULT_TIMEOUT);
    }

    /**
     * Retrieve device location.
     *
     * @return last known device location or null when it cannot be found
     */
    @Nullable
    public GeoCoordinates getLocation(long timeout) {
        GeoCoordinates coordinates = getCurrentLocation(timeout);
        if (coordinates == null) {
            coordinates = getLastLocation(timeout);
        }

        if (coordinates != null) {
            database.saveLastLocation(coordinates);
        } else {
            coordinates = database.getLastLocation();
            if (coordinates != null) {
                log.debug("Using location from local database");
            } else {
                log.error("Unable to obtain location from database");
            }
        }

        return coordinates;
    }

    @Nullable
    private GeoCoordinates getCurrentLocation(long timeout) {
        if (!isLocationPermissionsGranted(context)) {
            log.warn("Unable to read current location. Permission denied.");
            return null;
        }

        final AtomicReference<GeoCoordinates> coordinates = new AtomicReference<>();
        final CountDownLatch completeSignal = new CountDownLatch(1);

        final LocationRequest request = new LocationRequest()
                .setFastestInterval(1500)
                .setInterval(3000)
                .setPriority(PRIORITY_HIGH_ACCURACY);

        LocationCallback callback = new LocationCallback() {

            @Override
            public void onLocationResult(LocationResult result) {
                if (result != null) {
                    Location location = result.getLastLocation();
                    coordinates.set(new GeoCoordinates(location));

                    log.debug("Received current location: " + location);
                }
                completeSignal.countDown();
            }
        };

        client.requestLocationUpdates(request, callback, Looper.getMainLooper());
        awaitCompletion(completeSignal, timeout, "Current location request");
        client.removeLocationUpdates(callback);

        return coordinates.get();
    }

    @Nullable
    private GeoCoordinates getLastLocation(long timeout) {
        if (!isLocationPermissionsGranted(context)) {
            log.warn("Unable to read last location. Permission denied.");
            return null;
        }

        final AtomicReference<GeoCoordinates> coordinates = new AtomicReference<>();
        final CountDownLatch completeSignal = new CountDownLatch(1);

        client.getLastLocation().addOnSuccessListener(location -> {
            if (location != null) {
                coordinates.set(new GeoCoordinates(location));
                completeSignal.countDown();

                log.debug("Received last location: " + location);
            }
        });

        awaitCompletion(completeSignal, timeout, "Last location request");

        return coordinates.get();
    }

    private void awaitCompletion(CountDownLatch latch, long timeout, String requestName) {
        try {
            if (!latch.await(timeout, TimeUnit.MILLISECONDS)) {
                log.warn(requestName + " timeout expired");
            }
        } catch (InterruptedException x) {
            log.warn(requestName + " interrupted", x);
        }
    }

}

File: app/src/main/kotlin/com/bopr/android/smailer/GeoLocator.kt
package com.bopr.android.smailer

import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.Context
import android.location.Location
import android.os.Looper
import com.bopr.android.smailer.util.Mockable
import com.bopr.android.smailer.util.checkPermission
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices.getFusedLocationProviderClient
import org.slf4j.LoggerFactory
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Provides last device location.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@Mockable
class GeoLocator(private val context: Context, private val database: Database) {

    private val client: FusedLocationProviderClient = getFusedLocationProviderClient(context)

    /**
     * Retrieve device location.
     *
     * @return last known device location or null when it cannot be found
     */
    fun getLocation(): GeoCoordinates? = getLocation(DEFAULT_TIMEOUT.toLong())

    /**
     * Retrieve device location.
     *
     * @return last known device location or null when it cannot be found
     */
    fun getLocation(timeout: Long): GeoCoordinates? {
        var coordinates = getCurrentLocation(timeout)

        if (coordinates == null) {
            coordinates = getLastLocation(timeout)
        }

        if (coordinates == null) {
            coordinates = database.lastLocation
            if (coordinates != null) {
                log.debug("Using location from local database")
            } else {
                log.warn("Unable to obtain location from database")
            }
        } else {
            database.putLastLocation(coordinates)
        }

        return coordinates
    }

    private fun getCurrentLocation(timeout: Long): GeoCoordinates? {
        if (!isLocationPermissionsGranted(context)) {
            log.warn("Unable to read current location. Permission denied.")
            return null
        }

        var coordinates: GeoCoordinates? = null
        val latch = CountDownLatch(1)

        val callback: LocationCallback = object : LocationCallback() {

            override fun onLocationResult(result: LocationResult?) {
                result?.let {
                    val location = it.lastLocation

                    log.debug("Received current location: $location")
                    coordinates = GeoCoordinates(location)
                }
                latch.countDown()
            }
        }

        client.requestLocationUpdates(LocationRequest(), callback, Looper.getMainLooper())
        awaitLatch(latch, timeout, "Current location request")
        client.removeLocationUpdates(callback)

        return coordinates
    }

    private fun getLastLocation(timeout: Long): GeoCoordinates? {
        if (!isLocationPermissionsGranted(context)) {
            log.warn("Unable to read last location. Permission denied.")
            return null
        }

        var coordinates: GeoCoordinates? = null
        val latch = CountDownLatch(1)

        client.lastLocation.addOnSuccessListener { location: Location? ->
            location?.let {
                log.debug("Received last location: $location")

                coordinates = GeoCoordinates(location)
                latch.countDown()
            }
        }
        awaitLatch(latch, timeout, "Last location request")

        return coordinates
    }

    private fun awaitLatch(latch: CountDownLatch, timeout: Long, requestName: String) {
        try {
            if (!latch.await(timeout, TimeUnit.MILLISECONDS)) {
                log.warn("$requestName timeout expired")
            }
        } catch (x: InterruptedException) {
            log.warn("$requestName interrupted", x)
        }
    }

    companion object {

        private val log = LoggerFactory.getLogger("GeoLocator")
        private const val DEFAULT_TIMEOUT = 1000

        fun isLocationPermissionsGranted(context: Context): Boolean {
            return checkPermission(context, ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION)
        }
    }

}